src/utstring.h

```diff
@@ -121,6 +121,24 @@ do {                                                             \
 
 #define utstring_body(s) ((s)->d)
 
+#define utstring_adopt(s,b,l)                              \
+do {                                                       \
+  if (s) {                                                 \
+    (s)->d = (b);                                          \
+    (s)->n = (l);                                          \
+    (s)->i = (l);                                          \
```

Comment:
It needs a tweak. The other "put" functions (`utstring_printf`, `utstring_bincpy`) internally set the internal buffer length `n` one larger than the input size, to accommodate a trailing nul, and then forcibly put it there. (Look at the `utstring_bincpy` definition to see what I mean).
It's a convenience so there is a guaranteed nul terminator in the spare capacity part of the string. It is a quirk of utstring that I like because, it makes it compatible with printf (etc) that expect nul-terminated strings, while keeping the nul past the length `i` means that the content/length are unharmed.

One way to preserve that notion, is to switch `adopt` to a two parameter version `utstring_adopt(s,d)` that demands a C string as input, and then its implementation sets 
<pre>
s->i = strlen(d);
s->n = s->i +1 ;
</pre>

The problem with the current three-parameter adopt is that the caller can pass a non-terminated string, breaking the expectation that utstring always has a `nul` after the buffer. 

We can switch to the four-parameter version `utstring_adopt(s,d,i,n)` but, to preserve the `nul` expectation, I would explicitly force a nul at d[i] (which requires a preceding realloc if i==n).